pkg/util/docker/dockerfile/dockerfile_test.go

```diff
@@ -0,0 +1,278 @@
+		if err == nil {
+			t.Errorf("InsertInstructions: %s: got nil; want error", name)
+		}
+	}
+}
+
+// TestFindAll tests calling FindAll with multiple values of cmd.
+func TestFindAll(t *testing.T) {
+	instructions := `FROM scratch
+LABEL version=1.0
+FROM busybox
+ENV PATH=/bin
+`
+	node, err := parser.Parse(strings.NewReader(instructions))
+	if err != nil {
+		t.Fatalf("parse error: %v", err)
+	}
+	for cmd, want := range map[string][]int{
+		command.From:       {0, 2},
+		command.Label:      {1},
+		command.Env:        {3},
+		command.Maintainer: nil,
+		"UnknownCommand":   nil,
+	} {
+		got := FindAll(node, cmd)
+		if !reflect.DeepEqual(got, want) {
+			t.Errorf("FindAll(node, %q) = %#v; want %#v", cmd, got, want)
+		}
+	}
+}
+
+// TestFindAllNilNode tests calling FindAll with a nil *parser.Node.
+func TestFindAllNilNode(t *testing.T) {
+	cmd := command.From
+	got := FindAll(nil, cmd)
+	if got != nil {
+		t.Errorf("FindAll(nil, %q) = %#v; want nil", cmd, got)
+	}
+}
+
+// TestParseTreeToDockerfile tests calling ParseTreeToDockerfile with multiple
+// valid inputs.
+func TestParseTreeToDockerfile(t *testing.T) {
+	testCases := map[string]struct {
+		in   string
+		want string
+	}{
+		"empty input": {
+			in:   ``,
+			want: ``,
+		},
+		"only comments": {
+			in: `# This is a comment
+# and this is another comment
+	# while this is an indented comment`,
+			want: ``,
+		},
+		"simple Dockerfile": {
+			in: `FROM scratch
+LABEL version=1.0
+FROM busybox
+ENV PATH=/bin
+`,
+			want: `FROM scratch
+LABEL version=1.0
+FROM busybox
+ENV PATH=/bin
+`,
+		},
+		"Dockerfile with comments": {
+			in: `# This is a Dockerfile
+FROM scratch
+LABEL version=1.0
+# Here we start building a second image
+FROM busybox
+ENV PATH=/bin
+`,
+			want: `FROM scratch
+LABEL version=1.0
+FROM busybox
+ENV PATH=/bin
+`,
+		},
+	}
+	for name, tc := range testCases {
+		node, err := parser.Parse(strings.NewReader(tc.in))
+		if err != nil {
+			t.Errorf("%s: parse error: %v", name, err)
+			continue
+		}
+		got := ParseTreeToDockerfile(node)
+		want := []byte(tc.want)
+		if !reflect.DeepEqual(got, want) {
+			t.Errorf("ParseTreeToDockerfile: %s:\ngot:\n%swant:\n%s", name, got, want)
+		}
+	}
+}
+
+// TestParseTreeToDockerfileNilNode tests calling ParseTreeToDockerfile with a
+// nil *parser.Node.
+func TestParseTreeToDockerfileNilNode(t *testing.T) {
+	got := ParseTreeToDockerfile(nil)
+	if got != nil {
+		t.Errorf("ParseTreeToDockerfile(nil) = %#v; want nil", got)
+	}
+}
+
+// TestEnv tests calling Env with multiple inputs.
+func TestEnv(t *testing.T) {
+	testCases := []struct {
+		in   map[string]string
+		want string
+	}{
+		{
+			in:   nil,
+			want: ``,
+		},
+		{
+			in:   map[string]string{},
+			want: ``,
+		},
+		{
+			in: map[string]string{
+				"GOPATH": "/go",
+				"MSG":    "Hello World!",
+			},
+			want: `ENV "GOPATH"="/go" "MSG"="Hello World!"`,
+		},
+		{
+			in:   map[string]string{"☃": "'\" \\ / \b \f \n \r \t \x00"},
```

Comment:
But yeah, a corner case... to trigger that one would need to construct a BuildConfig with carefully chosen bytes, hope it goes all the way to the point those are to be added to your Dockerfile, and then instead of having a new image with a funny environment variable, you get an error from the Docker parser....
This code gives you the funny env vars :-)